### PR TITLE
Merge with latest `mlc-ai/main` (`5bc3ffa6f682a4cf42fdeba3a4c505d0e7c08c3c`)

### PR DESCRIPTION
**Pull request by sunggg**
This PR pulls the latest changes in the upstream. 
One of the key change is the renaming of `mlc_chat` to `mlc_llm` as they deprecate the legacy flow in pre-SLM era.
Once this is merged, I will follow-up with the ollm changes. 

During resolving the merge conflicts, I found that there are a couple of subtle differences in the quantization flow and presharding. 
I believe I currently kept the existing behavior, but it would be great if @csullivan @vinx13 @JosephTheOctonaut can confirm. 

For the review, it is safe to ignore other directories outside of `python/`.

### CODE REVIEW

File: python/mlc_llm/quantization/group_quantization.py [hunk 1]
```diff
@@ -128,6 +126,12 @@ def visit_module(self, name: str, node: nn.Module) -> Any:
                 """
                 from . import fp8_quantization as fp8
 
+                # TODO(@team): this is from upstream. commented out for now.
```
Comment (sunggg) [inline python/mlc_llm/quantization/group_quantization.py]: cc. @csullivan @JosephTheOctonaut @vinx13

File: python/mlc_llm/quantization/group_quantization.py [hunk 2]
```diff
@@ -983,3 +980,14 @@ def forward(self, x: nn.Tensor, indptr: nn.Tensor) -> nn.Tensor:  # pylint: disa
                 indptr_dtype=indptr.dtype,
                 group_size=self.group_size,
             )
+        # TODO(@team): this is from upstream. commented out for now.
```
Comment (sunggg) [inline python/mlc_llm/quantization/group_quantization.py]: cc. @csullivan @JosephTheOctonaut @vinx13

File: python/mlc_llm/quantization/group_quantization.py [hunk 3]
```diff
@@ -918,6 +914,7 @@ def from_mixtral_experts(
         if "shard_strategy" in src.weight.attrs:
             shard = src.weight.attrs["shard_strategy"]
             apply_sharding(shard, f"{shard.name}_q_weight", quantized_mistral_experts.q_weight)
+            # TODO(@team): upstream does not use this condition
```
Comment (sunggg) [inline python/mlc_llm/quantization/group_quantization.py]: cc. @csullivan @JosephTheOctonaut @vinx13

File: python/mlc_llm/quantization/group_quantization.py [hunk 4]
```diff
@@ -602,6 +597,7 @@ def from_linear(cls, src: nn.Linear, config: GroupQuantize) -> "GroupQuantizeLin
         if "shard_strategy" in src.weight.attrs:
             shard = src.weight.attrs["shard_strategy"]
             apply_sharding(shard, f"{shard.name}_q_weight", quantized_linear.q_weight)
+            # TODO(@team): upstream does not require this condition.
```
Comment (sunggg) [inline python/mlc_llm/quantization/group_quantization.py]: cc. @csullivan @JosephTheOctonaut @vinx13

File: python/mlc_llm/support/preshard.py [hunk 5]
```diff
@@ -101,35 +108,57 @@ def _compile_shard_funcs(mod: IRModule, device: Device):
...
     bb = relax.BlockBuilder()
     param_to_shard_func = {}
     shard_func_names = set()
-    for name, param in model.state_dict().items():
+    new_named_params: Dict[str, nn.Parameter] = {}
+    has_shard_strategy = False
+    for name, param in named_params.items():
         shard_strategy = param.attrs.get("shard_strategy", None)
         if shard_strategy is not None:
-            _update_quantize_map(quantize_map, named_params, name, tensor_parallel_shards)
```
Comment (vinx13) [inline python/mlc_llm/support/preshard.py:121]: need to keep original behavior because fp8 quantization depends on it

File: python/mlc_llm/loader/huggingface_loader.py [hunk 6]
```diff
@@ -115,13 +115,15 @@ def load(
         mlc_names = _loading_order(self.extern_param_map, self.torch_to_path)
         for mlc_name in tqdm(mlc_names):
             param = self._load_mlc_param(mlc_name, device=device)
-            if preshard_funcs is not None and mlc_name in preshard_funcs:
```
Comment (vinx13) [inline python/mlc_llm/loader/huggingface_loader.py]: keep this as well. The difference of this and the upstream is that, previously we applied presharding before quantization, each shard is quantized separately, this is not the expected behavior and it's also inconsistent with the behavior presharding is turned off, the upstream has fixed this behavior, it applied quantization first, and then presharding. However, current quantization depends on such behavior. We still need to keep this behavior for now

File: python/mlc_llm/support/preshard.py (hunk 5 above)
Comment (sunggg) [reply python/mlc_llm/support/preshard.py:121]: Done. thanks for the catch!

File: python/mlc_llm/loader/huggingface_loader.py (hunk 6 above)
Comment (sunggg) [reply python/mlc_llm/loader/huggingface_loader.py]: I see. Is someone working on this?

**Comment by sunggg** [comment]
@vinx13, thank you for the review. I think I reflected your comments, would you mind taking another look? Appreciate it!

File: python/mlc_llm/quantization/group_quantization.py (hunk 4 above)
Comment (vinx13) [reply python/mlc_llm/quantization/group_quantization.py]: we can remove it and all fp8 contents in this file

File: python/mlc_llm/quantization/group_quantization.py (hunk 3 above)
Comment (vinx13) [reply python/mlc_llm/quantization/group_quantization.py]: can remove this, it is also fp8

File: python/mlc_llm/quantization/group_quantization.py (hunk 2 above)
Comment (vinx13) [reply python/mlc_llm/quantization/group_quantization.py]: we can clean up fp8 here as well

File: python/mlc_llm/quantization/group_quantization.py (hunk 1 above)
Comment (vinx13) [reply python/mlc_llm/quantization/group_quantization.py]: moe_gate usually should be skipped or quantized specially, because moe gate is small and the impact on accuracy is larger
